### PR TITLE
feat: modified notFound() to throw new NotFound()

### PR DESCRIPTION
**Pull request by wyattjoh**
## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have a helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have a helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `pnpm lint`
- [ ] The "examples guidelines" are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing/examples/adding-examples.md)

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: fcb2495f66b658603d7cabe64c2270917efbd27e

`pnpm testheadless test/integration/relay-analytics/test/index.test.js` 
- Analytics relayer with exported method > reports attribution

<details>
<summary>Expand output</summary>

  ● Analytics relayer with exported method › reports attribution

    TypeError: Cannot read properties of null (reading 'find')

      141 |     )
      142 |     const metrics = JSON.parse(str)
    > 143 |     const LCP = metrics.find((m) => m.name === 'LCP')
          |                         ^
      144 |     expect(LCP).toBeDefined()
      145 |     expect(LCP.attribution).toBeDefined()
      146 |     expect(LCP.attribution.element).toBe('#__next>div>h1')

      at Object.<anonymous> (integration/relay-analytics/test/index.test.js:143:25)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/client/components/not-found.ts [hunk 1]
```diff
@@ -1,8 +1,5 @@
 export const NOT_FOUND_ERROR_CODE = 'NEXT_NOT_FOUND'
 
-export function notFound() {
-  // eslint-disable-next-line no-throw-literal
-  const error = new Error(NOT_FOUND_ERROR_CODE)
-  ;(error as any).digest = NOT_FOUND_ERROR_CODE
-  throw error
+export class NotFound extends Error {
```
Comment (sebmarkbage) [inline packages/next/client/components/not-found.ts:3]: This doesn't work if transpiled.

**Comment by sebmarkbage** [comment]
What's this for?

**Comment by wyattjoh** [comment]
Was briefly exploring an alternative syntax of:

```tsx
import { NotFound } from 'next/navigation';

const User = async () => {
  const user = await getUser();
  if (!user) {
    throw new NotFound();
  }

  return <div>{user.name}</div>;
}
```

Instead of:

```tsx
import { notFound } from 'next/navigation';

const User = async () => {
  const user = await getUser();
  if (!user) {
    throw notFound();
  }

  return <div>{user.name}</div>;
}
```

Because technically, `notFound()` already throws, I thought it might be more ergonomic to construct the error directly and throw that. This also avoids the typescript errors we see when instead of throwing, we return, as `void` isn't a valid return type from a component.

Open to ideas to resolve this, but as it stands, I don't feel as though it's a very good developer experience.

**Comment by sebmarkbage** [comment]
I believe we can type `notFound()` and `redirect()` with the return type `: never`.

That way this should work without type errors:

```tsx
import { notFound } from 'next/navigation';

const User = async () => {
  const user = await getUser();
  if (!user) {
    notFound();
  }

  return <div>{user.name}</div>;
```

That way you don't have to return nor throw.

**Comment by timneutkens** [comment]
The `: never` change was implemented.

**Pull request closed without merging**